pep-9999.rst

```diff
@@ -0,0 +1,1245 @@
+                                       adds a marker file)
+3    Distribution Python in venv       Currently yes; stays yes    There are no externally-installed packages
+4    Distribution Python in venv       Currently yes; stays yes    Currently no; stays no
+     with ``--system-site-packages``
+5    Distribution Python in Docker     Currently yes; stays yes    Currently yes; becomes no
+                                       (assuming the Docker image
+                                       removes the marker file)
+6    Conda environment                 Currently yes; stays yes    Currently yes; stays yes
+7    Dev-facing distribution           Currently yes; becomes no   Currently often yes; becomes no
+                                       (assuming they add a        (assuming they configure ``sysconfig`` as needed)
+                                       marker file)
+8    Distribution building packages    Currently yes; can stay yes Currently yes; becomes no
+9    ``PYTHONHOME`` copied from        Currently yes; becomes no   Currently yes; becomes no
+     a distribution Python stdlib
+10   ``PYTHONHOME`` copied from        Currently yes; stays yes    Currently yes; stays yes
+     upstream Python stdlib
+==== ================================= =========================== ===================================================
+
+In more detail, the use cases above are:
+
+1. A standard unpatched CPython, without any special
+   configuration of or patches to ``sysconfig`` and without
+   a marker file. This PEP does not change its behavior.
+
+   Such a CPython should (regardless of this PEP) not be
+   installed in a way that that overlaps any
+   distribution-installed Python on the same system. For
+   instance, on an OS that ships Python in ``/usr/bin``, you
+   should not install a custom CPython built with
+   ``./configure --prefix=/usr``, or it will overwrite some
+   files from the distribution and the distribution will
+   eventually overwrite some files from your installation.
+   Instead, your installation should be in a separate
+   directory (perhaps ``/usr/local``, ``/opt``, or your home
+   directory).
+
+   Therefore, we can assume that such a CPython has its own
+   ``stdlib`` directory and its own ``sysconfig`` schemes
+   that do not overlap any distribution-installed Python. So
+   any OS-installed packages are not visible or relevant
+   here.
+
+   If there is a concept of "externally-installed" packages
+   in this case, it's something outside the OS and generally
+   managed by whoever built and installed this CPython.
+   Because the installer chose not to add a marker file or
+   modify ``sysconfig`` schemes, they're choosing the
+   current behavior, and ``pip install`` can remove any
+   packages available in this CPython.
+
+2. A distribution's ``/usr/bin/python3``, either when
+   running ``pip install`` as root or ``pip install
+   --user``, following our `Recommendations for
+   distributions`_.
```

Comment:
Question: is there a way to limit `pip install` while still allowing `pip install --user`? I've read the whole PEP (in 1-2 sittings) and didn't see anything explicit about it.

My main context is the Windows Store package of the python.org distro, which has a very locked down system site-packages (containing packages that would otherwise be upgradeable), but also sumo distros in general (which I guess is kind of similar to the monorepo case). If I've got a standard distro including packages that are preferred, but still want to allow individual users to override (not remove/replace) them, I'd love to have the mechanism be standardised.

I'm guessing bundling the user packages in with site packages is because none of the distros currently use the `-s` option when invoking their own commands?

Reply:
`pip install --user` can still cause these issues. Here's an example bug I was looking at, today, caused by a user not realising what they'd installed to `--user`: https://bugs.launchpad.net/bugs/1912535

> My main context is the Windows Store package of the python.org distro

I wouldn't expect that to set EXTERNALLY-MANAGED, because they don't provide Python libraries in the Windows Store (do they?). So, they shouldn't be concerned about pip installing things. Even if file permissions mean that `--user` is the only way pip can install, that shouldn't be an issue.

Reply:
> Here's an example bug I was looking at, today, caused by a user not realising what they'd installed to --user

Or arguably, caused a system-installed tool _not_ passing `-s` to ignore user-installed packages?

> I wouldn't expect that to set EXTERNALLY-MANAGED, because they don't provide Python libraries in the Windows Store (do they?)

We could, though the amount of overhead isn't really worth it (at least at the ridiculously high rate of releases everything has). And yeah, the super-restrictive permissions are going to break anyone who tries to install into the default directory anyway. Pip already has fallback behaviour, but everyone else will figure it out eventually after they get bug reports (unless there's a spec that they're already following to detect that they shouldn't even attempt to use that folder and to display a message explaining why...)

Seeing as the `EXTERNALLY-MANAGED` file can already contain options, why not have an option for `managed-schemes: posix_user, nt_user` or similar? Make it default to what's already specified in the PEP, and then it can be overridden for fewer/greater restrictions?

Reply:
> Or arguably, caused a system-installed tool not passing -s to ignore user-installed packages?

Nothing specifies `-s`. Or at least, I've never seen anything that does.

Reply:
Petr posted on linux-sig that Red Hat is moving to start using `-s` everywhere that they want to avoid conflicts with user-specified packages. So apparently it's possible for distros to change here, which would enable user packages to be used as intended.

Reply:
Right - if the goal here is "pip shouldn't install systemwide because the user doesn't have permissions" as opposed to "pip shouldn't install systemwide because it will risk breaking software shipped expecting a particular Python," then I think the thing most in line with the worldview of this PEP is to set your default sysconfig scheme to be `nt_user` or something, so that that's where `pip install` installs things to, and don't configure anything else. Don't set `EXTERNALLY-MANAGED`. You can still ship software with that installation of Python in the `nt` scheme and it will get loaded, and because of the second change of this PEP, pip won't remove them if it installs a new version of something to `nt_user`.

(I actually think it already won't remove them, but I'm not positive. Also note that I haven't tested that setting the default sysconfig scheme to a user scheme causes pip to do what you expect, but it seems like it ought to / this PEP sort of assumes that installers use the default sysconfig scheme by default.)

`EXTERNALLY-MANAGED` is for when you're simultaneously shipping Python as a thing for end users to use directly and also as an implementation language for software that you want to not be fragile. If I understand your use case correctly, there's no software using the Windows Store distribution of Python as an interpreter (other than Python-specific stuff like pip, idle, pydoc, etc.), right?

Or put another way, I think your goal is not to prevent users adding things to the global (virtualenv-less) `sys.path` and breaking things, because there's nothing they would break; your goal is simply to prevent installers from defaulting to the global installation directory.

But I don't know a ton about this use case and might be misunderstanding it.

None of the distros currently use `-s` when invoking their own commands because there are legitimate reasons to add software to the global context (e.g., running /usr/bin/ansible but installing Ansible extensions / writing custom Python). The position of this PEP is that adding software to the global context should take explicit user action but it should, generally, be permitted (or at least that distros should be able to opt particular applications into supporting it).

Reply:
I guess I'm still not getting how you can allow users to install with `--user` by preventing them from installing with `--user`. Is the plan to teach everyone to just add another option to be able to install things "globally"? Because that doesn't sound like much of an improvement over the current situation.

> or at least that distros should be able to opt particular applications into supporting it

The officially supported upstream way of doing this on a per-application level is the `-s` option. So you would default every application to use `-s`, and then those that want to use user-installed extensions should omit it. Adding (or removing) the user site packages path directly is also fairly straightforward if you control the application (and e.g. want a CLI option for "safe mode" that ignores user packages).

Basically what I'm getting at is that I totally agree with an interpreter-level instruction to package managers "you should not be modifying this directory", but I think we're pushing it too far and forcing legitimate, supported scenarios into needing workarounds. We don't have to push it too far.

Reply:
> I guess I'm still not getting how you can allow users to install with --user by preventing them from installing with --user.

The goal is to make them jump through a hoop if they want to install with `--user`. Don't do it by default, but make them explicitly opt-in. And realise what they're opting into, e.g. suggest alternatives that are better for what they're trying to do.

`sudo` was not a sufficient opt-in for doing things globally that can break your system, because people are used to sudo enough that they just do it without thinking.

Reply:
Why won't that happen again with the new flag? Are we relying on Python being enough of an edge case that people won't internalise (or write blog posts about) the new way?

Reply:
Because we'll be suggesting mechanisms that actually work, e.g. pipx and virtualenvs, rather than just saying if you really want to do this, use a bigger hammer.

Reply:
Those aren't going to work for adding an extension to system-installed Ansible though, right? (to pick up on the example given above)

Reply:
Right, that's the situation where you need to pass a flag to say, yes I really mean to do this.